text/0074-alphanumeric-transfer-ids.md

```diff
@@ -0,0 +1,130 @@
+#[cfg_attr(feature = "json-schema", derive(JsonSchema))]
+pub enum Memo {
+    None,
+    Numeric(u64),
+    Alphanumeric(String),
+}
+
+impl Memo {
+    fn tag(&self) -> u8 {
+        match self {
+            Memo::None => bytesrepr::OPTION_NONE_TAG,
+            Memo::Numeric(_) => bytesrepr::OPTION_SOME_TAG,
+            Memo::Alphanumeric(_) => 2,
+        }
+    }
+}
+
+impl Default for Memo {
+    fn default() -> Self {
+        Memo::None
+    }
+}
+
+impl ToBytes for Memo {
+    fn to_bytes(&self) -> Result<Vec<u8>, bytesrepr::Error> {
+        let mut bytes = vec![self.tag()];
+        match self {
+            Memo::None => (),
+            Memo::Numeric(value) => {
+                bytes.extend(value.to_bytes()?);
+            }
+            Memo::Alphanumeric(value) => {
+                bytes.extend(value.to_bytes()?);
+            }
+        }
+        Ok(bytes)
+    }
+
+    fn serialized_length(&self) -> usize {
+        1 + match self {
+            Memo::None => 0,
+            Memo::Numeric(value) => value.serialized_length(),
+            Memo::Alphanumeric(value) => value.serialized_length(),
+        }
+    }
+}
+
+impl FromBytes for Memo {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), bytesrepr::Error> {
+        let (tag, remainder) = u8::from_bytes(bytes)?;
+        match tag {
+            bytesrepr::OPTION_NONE_TAG => Ok((Memo::None, remainder)),
+            bytesrepr::OPTION_SOME_TAG => {
+                let (value, remainder) = u64::from_bytes(remainder)?;
+                Ok((Memo::Numeric(value), remainder))
+            }
+            2 => {
+                let (value, remainder) = String::from_bytes(remainder)?;
+                Ok((Memo::Alphanumeric(value), remainder))
+            }
+            _ => Err(bytesrepr::Error::Formatting),
+        }
+    }
+}
+```
+
+`Transfer` would then be changed to hold a non-optional `Memo` instead of the current `Option<u64>`. This relies on the way that we're currently serializing `Option<64>`. The tag for `Option::None` corresponds to `Memo::None` and the `Option::Some` tag corresponds to `Memo::Numeric`. The tag for `Memo::Alphanumeric` is new and will not conflict with the existing tags.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+Major changes to core features of the blockchain network require significant time and effort to implement, and may also come with unforeseen consequences.
+
+## Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The `Memo` type can be implemented without requiring any data migrations because of the way that the current `Option` tag bytes correspond to the new `Memo` tag bytes. Adding a new, independent text field to `Transaction` would require a production data migration, and it could be confusing to users that are trying to understand the difference between the optional integral type `id` field and the alphanumeric memo field.
```

Comment:
Personally, I think that the `Memo` field should be added to a _deploy_ type and not `Transfer` variant. All transers are deploys but not all deploys are transfers. By adding the `Memo` to transfer only we're limiting users' capabilities. It's quite a common practice to attach memo to a contract call on other networks. For example, in Terra it's [part of transaction body](https://github.com/terra-money/terra.js/blob/9e5f553de3ff3e975eaaf91b1f06e45658b1a5e0/src/client/lcd/api/TxAPI.ts#L247).